### PR TITLE
hold single AudioContext instance

### LINKED ISSUE #1694: Exception when bell sound is used
Repro:

1. Enable bell in vscode
2. Open a terminal
3. Try trigger bell

```
Uncaught DOMException: Failed to construct 'AudioContext': The number of hardware contexts provided (6) is greater than or equal to the maximum bound (6).
    at SoundManager.playBellSound (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/SoundManager.js:11:34)
    at Terminal.bell (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/Terminal.js:1035:31)
    at InputHandler.bell (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/InputHandler.js:299:24)
    at /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/InputHandler.js:131:94
    at EscapeSequenceParser.parse (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/EscapeSequenceParser.js:310:25)
    at InputHandler.parse (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/InputHandler.js:200:22)
    at Terminal._innerWrite (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/Terminal.js:876:32)
    at /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/node_modules.asar/vscode-xterm/lib/Terminal.js:858:23
```

screen shot 2018-09-15 at 9 42 01 am

### PR DESCRIPTION
**Pull request by jerch**
Fixes #1694.

It should be save to just hold one `AudioContext` instance for all terminal instances within a document (created on the first bell request on `SoundManager`).

NB: ~The change will not work if there are still more audio contexts spawned than allowed on integrator side (e.g. vscode). If thats an issue we might want to refactor SoundManager to take an optional argument `context`. This way the integrator can maintain the audio contexts itself.~ Added optional ctor argument.

### CODE REVIEW

**Comment by jerch** [comment]
Up for review.

File: src/SoundManager.ts [hunk 1]
```diff
@@ -12,30 +12,37 @@ import { ITerminal, ISoundManager } from './Types';
...
+      }
+      this._audioContext = new audioContextCtor();
+    }
+    return SoundManager._audioContext;
+  }
 
   constructor(
-    private _terminal: ITerminal
+    private _terminal: ITerminal,
+    private _audioContext?: AudioContext
```
[the remark this answers is not in the document]
Comment (jerch) [reply src/SoundManager.ts]: You mean the optional audio context arg?

[the remark this answers is not in the document]
Comment (Tyriar) [reply src/SoundManager.ts]: Yeah, we could always add it if we wanted to expose it. I'd say the likelihood of the context limit being hit in a terminal app is pretty low.

[the remark this answers is not in the document]
Comment (jerch) [reply src/SoundManager.ts]: Right, and carrying this through up to xterm's entry API seems silly.

**Comment by jerch** [comment]
Ok done.

**Pull request merged**